### PR TITLE
YLP-339 jwt token for Zendesk and other external services

### PR DESCRIPTION
**Pull request by ydrisR**
The goal of this change is to add a new route "/sso/{service}" which generates a new token signed with a key dedicated to a 3rd party service. This way we can use shoreline to authenticate users to other services without sharing the internal secret. 
This new route is only available for authenticated users.

### CODE REVIEW

**Comment by BigFab** [comment]
would that not work to have the token for external services be created and attached to the user when he is logging in?

**Comment by BigFab** [comment]
Brilliant!
I just have a doubt on whether we should expose the userid in the token for external services. I think we should not. Here it should not be needed for Zendesk

File: user/api.go [hunk 1]
```diff
@@ -199,9 +200,10 @@ const (
 	//api logging prefix
 	USER_API_PREFIX = "api/user "
 
-	TP_SERVER_NAME   = "x-tidepool-server-name"
-	TP_SERVER_SECRET = "x-tidepool-server-secret"
-	TP_SESSION_TOKEN = "x-tidepool-session-token"
+	TP_SERVER_NAME    = "x-tidepool-server-name"
+	TP_SERVER_SECRET  = "x-tidepool-server-secret"
+	TP_SESSION_TOKEN  = "x-tidepool-session-token"
+	EXT_SESSION_TOKEN = "x-external-session-token"
```
Comment (BigFab) [inline user/api.go:206]: x-sso-session-token?

File: user/api.go [hunk 2]
```diff
@@ -293,6 +295,8 @@ func (a *Api) SetHandlers(prefix string, rtr *mux.Router) {
 
 	rtr.Handle("/token/{token}", varsHandler(a.ServerCheckToken)).Methods("GET")
 
+	rtr.Handle("/sso/{service}", varsHandler(a.Get3rdPartyToken)).Methods("POST")
```
Comment (BigFab) [inline user/api.go]: actually I have a doubt on the name here, as this is not true SSO. Shall we name it differently? ext_token?

File: user/api.go [hunk 3]
```diff
@@ -998,6 +1001,60 @@ func (a *Api) AnonymousIdHashPair(res http.ResponseWriter, req *http.Request) {
...
+	}
+
+	if secret == "" {
+		// the secret is not defined for this service
+		a.logger.Println(http.StatusBadRequest, "the service does not exist")
+		sendModelAsResWithStatus(res, status.NewStatus(http.StatusBadRequest, STATUS_ERR_GENERATING_TOKEN), http.StatusBadRequest)
+		return
+	}
+
+	// Should we also support authentication with login/pwd?
```
Comment (BigFab) [inline user/api.go]: probably when we have a login form?

File: user/token_test.go [hunk 4]
```diff
@@ -1,247 +1,326 @@
...
+	}
+
+	td, _ := UnpackSessionTokenAndVerify(token.ID, tokenConfig.Secret)
+
+	if td.DurationSecs != tokenConfig.DurationSecs {
+		t.Fatalf("the duration should be from config")
+	}
+}
+
+func Test_GenerateSessionToken_DurationSecsTrumpConfig(t *testing.T) {
```
Comment (BigFab) [inline user/token_test.go:83]: il vient faire quoi ici le canard ?

File: user/token_test.go [hunk 5]
```diff
@@ -1,247 +1,326 @@
...
+
+	jwtToken, err := jwt.Parse(token.ID, func(t *jwt.Token) (interface{}, error) { return []byte(testData.config.Secret), nil })
+	if err != nil || !jwtToken.Valid {
+		t.Fatalf("should decode and validate the token")
+	}
+
+	claims := jwtToken.Claims.(jwt.MapClaims)
+
+	// check zendesk claims
+	if claims["organization"] != "Patient" {
```
Comment (BigFab) [inline user/token_test.go:125]: should we also check the duration? iat? subject? exp?

Comment (BigFab) [reply user/token_test.go:125]: or it it part of the generic session token?

**Review by BigFab — CHANGES_REQUESTED** [review]

File: user/token_test.go (hunk 4 above)
Comment (ydrisR) [reply user/token_test.go:83]: 🦆 ?

File: user/token_test.go (hunk 5 above)
Comment (ydrisR) [reply user/token_test.go:125]: I added this in the generic part

**Review by BigFab — APPROVED** [review]
LGTM

**Pull request merged**